genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java

```diff
@@ -331,10 +342,81 @@ private void setRuntimeEnvironment(
     private Cluster getCluster(final JobRequest jobRequest) throws GenieException {
         final long start = System.nanoTime();
         try {
-            log.info("Selecting cluster for job {}", jobRequest.getId());
-            final Cluster cluster
-                = this.clusterLoadBalancer.selectCluster(this.clusterService.chooseClusterForJobRequest(jobRequest));
-            log.info("Selected cluster {} for job {}", cluster.getId(), jobRequest.getId());
+            log.info("Selecting cluster for job {}", jobRequest.getId().orElse(NO_ID_FOUND));
+            final List<Cluster> clusters = this.clusterService.chooseClusterForJobRequest(jobRequest);
+            Cluster cluster = null;
+            if (clusters.isEmpty()) {
+                throw new GeniePreconditionException(
+                    "No cluster/command combination found for the given criteria. Unable to continue"
+                );
+            } else if (clusters.size() == 1) {
+                cluster = clusters.get(0);
+            } else {
+                for (final ClusterLoadBalancer loadBalancer : this.clusterLoadBalancers) {
+                    final String loadBalancerClass = loadBalancer.getClass().getCanonicalName();
+                    try {
+                        final Cluster selectedCluster = loadBalancer.selectCluster(clusters, jobRequest);
+                        if (selectedCluster != null) {
+                            log.debug(
+                                "Successfully selected cluster {} using load balancer {}",
+                                selectedCluster.getId().orElse(NO_ID_FOUND),
+                                loadBalancerClass
+                            );
+                            this.registry.counter(
+                                this.loadBalancerId
+                                    .withTag(
+                                        LOAD_BALANCER_CLASS_TAG,
+                                        loadBalancerClass
+                                    )
+                                    .withTag(
+                                        LOAD_BALANCER_STATUS_TAG,
+                                        LOAD_BALANCER_STATUS_SUCCESS
+                                    )
+                            ).increment();
+                            cluster = selectedCluster;
+                            break;
+                        } else {
+                            this.registry.counter(
+                                this.loadBalancerId
+                                    .withTag(
+                                        LOAD_BALANCER_CLASS_TAG,
+                                        loadBalancerClass
+                                    )
+                                    .withTag(
+                                        LOAD_BALANCER_STATUS_TAG,
+                                        LOAD_BALANCER_STATUS_FAILURE
```

Comment:
Here returning null is considered FAILURE in terms of metrics, but maybe something like "NO_PREFERENCE" is more accurate?